task-graphs-v2.yaml

```diff
@@ -0,0 +1,370 @@
+
+          For Python (and most other languages), all positional arguments will
+          come before all named arguments (if any are present):
+
+              // fn(arg1, arg2, arg3)
+              [
+                {value: arg1},
+                {value: arg2},
+                {value: arg3},
+              ]
+              // fn(arg1, arg2, n=kw1, a=kw2)
+              [
+                {value: arg1},
+                {value: arg2},
+                {name: "n", value: kw1},
+                {name: "a", value: kw2},
+              ]
+              // fn(kw=k1, only=k2)
+              [
+                {name: "kw", value: k1},
+                {name: "only", value: k2},
+              ]
+
+          However, in R, named and positional arguments may be intermixed
+          freely:
+
+              // fn(arg, n=kw1, arg2)
+              [
+                {value: arg},
+                {name: "n", value: kw1},
+                {value: arg2},
+              ]
+        type: array
+        items:
+          $ref: '#/definitions/TGUDFArgument'
+
+  TGUDFArgument:
+    description: >
+      A single argument to a UDF. This may represent a positional argument or
+      a named argument, depending upon whether `name` is set.
+    type: object
+    properties:
+      name:
+        description: The name of the argument, if present.
+        type: string
+        x-nullable: true
+      value:
+        $ref: '#/definitions/TGArgument'
+
+  TGUDFEnvironment:
+    description: Metadata about the environment where we want to execute a UDF.
+    type: object
+    properties:
+      language:
+        $ref: '#/definitions/UDFLanguage'
+      image_name:
+        description: >
+          The name of the image to use for the execution environment.
+
+
+  #
+  # Parameters
+  #
+
+  TGArgument:
+    description: >
+      An argument provided to a node. This is one of a direct value (i.e.,
+      a raw JSON value) or a `TGSentinel`.
+
+      For example this Python value:
+
+          {"a": [1, "pipe", range(30), None], "b": b"bytes"}
+
+      is encoded thusly (with included comments):
+
+          {  // A dictionary with string keys is JSON-encodable.
+            "a": [  // As is a list.
+              1,
+              "pipe",
+              {"__tiledb_sentinel__": {  // A `range` is replaced with its pickle.
+                "immediate": {
+                  "format": "python_pickle",
+                  "base64_data": "gASVIAAAAAAAAACMCGJ1aWx0aW5zlIwFcmFuZ2WUk5RLAEseSwGHlFKULg=="
+                }
+              }},
+              null
+            ],
+            "b": {"__tiledb_sentinel__": {  // Raw binary data is encoded into base64.
+              "immediate": {
+                "format": "bytes",
+                "base64_data": "Ynl0ZXM="
+              }
+            }}
+          }
+
+
+  # The below messages are not used directly in any API definitions,
+  # but are included as documentation of data encodings.
+
+  TGSentinel:
```

Comment:
Naming. Sentinel is counter intuitive here. Some suggestions `Slot` `FormalArgument` `Placeholder` `Tap` no strong opinion however